flink/src/main/java/org/apache/iceberg/flink/FlinkDynamicTableFactory.java

```diff
@@ -19,56 +19,155 @@
+import org.apache.flink.table.catalog.ObjectIdentifier;
 import org.apache.flink.table.catalog.ObjectPath;
+import org.apache.flink.table.catalog.exceptions.CatalogException;
+import org.apache.flink.table.catalog.exceptions.DatabaseAlreadyExistException;
+import org.apache.flink.table.catalog.exceptions.TableAlreadyExistException;
 import org.apache.flink.table.connector.sink.DynamicTableSink;
 import org.apache.flink.table.connector.source.DynamicTableSource;
 import org.apache.flink.table.factories.DynamicTableSinkFactory;
 import org.apache.flink.table.factories.DynamicTableSourceFactory;
 import org.apache.flink.table.utils.TableSchemaUtils;
+import org.apache.flink.util.Preconditions;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
 
 public class FlinkDynamicTableFactory implements DynamicTableSinkFactory, DynamicTableSourceFactory {
+  private static final String FACTORY_IDENTIFIER = "iceberg";
+
+  private static final ConfigOption<String> CATALOG_NAME =
+      ConfigOptions.key("catalog-name")
+          .stringType()
+          .noDefaultValue()
+          .withDescription("Catalog name");
+
+  private static final ConfigOption<String> CATALOG_TYPE =
+      ConfigOptions.key(FlinkCatalogFactory.ICEBERG_CATALOG_TYPE)
+          .stringType()
+          .noDefaultValue()
+          .withDescription("Catalog type, the optional types are: custom, hadoop, hive.");
+
+  private static final ConfigOption<String> CATALOG_DATABASE =
+      ConfigOptions.key("catalog-database")
+          .stringType()
+          .defaultValue(FlinkCatalogFactory.DEFAULT_DATABASE_NAME)
+          .withDescription("Database name that managed in the iceberg catalog.");
+
   private final FlinkCatalog catalog;
 
+  public FlinkDynamicTableFactory() {
+    this.catalog = null;
+  }
+
   public FlinkDynamicTableFactory(FlinkCatalog catalog) {
     this.catalog = catalog;
   }
 
   @Override
   public DynamicTableSource createDynamicTableSource(Context context) {
-    ObjectPath objectPath = context.getObjectIdentifier().toObjectPath();
-    TableLoader tableLoader = createTableLoader(objectPath);
+    ObjectIdentifier objectIdentifier = context.getObjectIdentifier();
+    Map<String, String> tableProps = context.getCatalogTable().getOptions();
+    CatalogTable catalogTable = context.getCatalogTable();
     TableSchema tableSchema = TableSchemaUtils.getPhysicalSchema(context.getCatalogTable().getSchema());
-    return new IcebergTableSource(tableLoader, tableSchema, context.getCatalogTable().getOptions(),
-        context.getConfiguration());
+
+    TableLoader tableLoader;
+    if (catalog != null) {
+      tableLoader = createTableLoader(catalog, objectIdentifier.toObjectPath());
+    } else {
+      tableLoader = createTableLoader(catalogTable, tableProps, objectIdentifier.getObjectName());
```

Comment:
Does this discard the rest of the identifier?

Reply:
I think this comment is similar to the @stevenzwu 's [question](https://github.com/apache/iceberg/pull/2666#discussion_r647120788) .   Saying if the flink table identifier is `flink_catalog.flink_database.table_name`  (when creating table using the `connector=iceberg`, not the iceberg catalog approach), then it's mapping the flink table with name `table_name` to the underlying iceberg table `table_name` with the configured database in table property.  The flink's `flink_catalog` and `flink_database` does not has any relationship to the iceberg's catalog & database. 

Maybe it's better to forbidden people to specify a `connector=iceberg` property when creating table under the iceberg catalog  ( adding a Precondition.checkArgument in `createDynamicTableSource` and `createDynamicTableSink`),  then we don't mixed the two approaches to create flink+iceberg table or map flink table to iceberg table.

Reply:
Okay, I thought that this was a way to run DDL even if you don't have an Iceberg catalog defined. It sort of does that, but it also creates a reference in the in-memory catalog. That's fine, but it does bring up a couple other questions:
* How do you create an in-memory catalog table pointing to an Iceberg table if the Iceberg table already exists? What if the DDL, like the schema, doesn't match?
* Why share the table name between the in-memory and external catalog but not the database? I think it makes sense to default the external database and table name using the ones from the DDL command but allow both to be overridden.

Reply:
> How do you create an in-memory catalog table pointing to an Iceberg table if the Iceberg table already exists? What if the DDL, like the schema, doesn't match?

If the underlying iceberg table already exists, the we still need to create the in-memory catalog table pointing to it.  If the in-memory catalog table schema does not match the underlying iceberg table, then the create table statement won't throw any exception but when executing the `SELECT` query or `INSERT INTO` query it will throw exception if the RowData read from iceberg table could not be parsed by the in-memory catalog table schema.   People will need to re-create the in-memory table and map to the underlying table once again. That's the default behavior for flink users,  because almost of the flink connectors are the similar behavior ( such as [JDBC](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/connectors/table/jdbc/) connector, [hive](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/connectors/table/hive/hive_read_write/) connector, [hbase](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/connectors/table/hbase/) connector).

Reply:
> Why share the table name between the in-memory and external catalog but not the database?

I think you are right.  I checked the [jdbc](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/connectors/table/jdbc/) connector,  we  could specify a different jdbc table name when creating the flink table.   It make sense to allow both db & table name to be overridden.